pkg/rotator/rotator.go

```diff
@@ -0,0 +1,536 @@
+package rotator
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	certName               = "tls.crt"
+	keyName                = "tls.key"
+	caCertName             = "ca.crt"
+	caKeyName              = "ca.key"
+	rotationCheckFrequency = 12 * time.Hour
+	certValidityDuration   = 10 * 365 * 24 * time.Hour
+)
+
+var crLog = logf.Log.WithName("cert-rotation")
+
+var vwhGVK = schema.GroupVersionKind{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "ValidatingWebhookConfiguration"}
+
+var _ manager.Runnable = &CertRotator{}
+
+// AddRotator adds the CertRotator and ReconcileVWH to the manager.
+func AddRotator(mgr manager.Manager, cr *CertRotator, vwhName string) error {
+	cr.client = mgr.GetClient()
+	cr.certsNotMounted = make(chan struct{})
+	if err := mgr.Add(cr); err != nil {
+		return err
+	}
+
+	vwhKey := types.NamespacedName{Name: vwhName}
+	reconciler := &ReconcileVWH{
+		client:    mgr.GetClient(),
+		scheme:    mgr.GetScheme(),
+		ctx:       context.Background(),
+		secretKey: cr.SecretKey,
+		vwhKey:    vwhKey,
+	}
+	if err := addController(mgr, reconciler, cr.SecretKey, vwhKey); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CertRotator contains cert artifacts and a channel to close when the certs are ready.
+type CertRotator struct {
+	client          client.Client
+	SecretKey       types.NamespacedName
+	CertDir         string
+	CAName          string
+	CAOrganization  string
+	DNSName         string
+	CertsMounted    chan struct{}
+	certsNotMounted chan struct{}
+}
+
+// Start starts the CertRotator runnable to rotate certs and ensure the certs are ready.
+func (cr *CertRotator) Start(stop <-chan (struct{})) error {
+	// explicitly rotate on the first round so that the certificate
+	// can be bootstrapped, otherwise manager exits before a cert can be written
+	crLog.Info("starting cert rotator controller")
+	defer crLog.Info("stopping cert rotator controller")
+	if err := cr.refreshCertIfNeeded(); err != nil {
```

Comment:
Just trying to make sure I understand the implications here: I think this is a blocking call; is that right? But the ensureCertsExist call is async? If so, these two calls seem at odds with each other?

Reply:
This is just what's checked in now; I don't know the code well myself. My personal preference would be to leave the first checkin as-is (I would have been happy even if it didn't compile for history purposes) and then file issues for all these concerns. wdyt?